internal/middleware/perfinsights/perfinsights.go

```diff
@@ -0,0 +1,189 @@
+	ResourceRelationLabel APIShapeLabel = "resource_relation"
+	SubjectTypeLabel      APIShapeLabel = "subject_type"
+	SubjectRelationLabel  APIShapeLabel = "subject_relation"
+	NameLabel             APIShapeLabel = "name"
+	FilterLabel           APIShapeLabel = "filter"
+)
+
+var allLabels = []string{
+	string(ResourceTypeLabel),
+	string(ResourceRelationLabel),
+	string(SubjectTypeLabel),
+	string(SubjectRelationLabel),
+	string(NameLabel),
+	string(FilterLabel),
+}
+
+// APIShapeLabels is a map of APIShapeLabel to any value.
+type APIShapeLabels map[APIShapeLabel]any
+
+// NoLabels returns an empty APIShapeLabels map.
+func NoLabels() APIShapeLabels {
+	return APIShapeLabels{}
+}
+
+// APIShapeLatency is the metric that SpiceDB uses to keep track of the latency
+// of API calls, by shape.
+var APIShapeLatency = promauto.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace:                   "spicedb",
+	Subsystem:                   "perf_insights",
+	Name:                        "api_shape_latency_seconds",
+	Help:                        "The latency of API calls, by shape",
+	Buckets:                     nil,
+	NativeHistogramBucketFactor: 1.1,
+}, append([]string{"api_kind"}, allLabels...))
+
+// ShapeBuilder is a function that returns a slice of strings representing the shape of the API call.
+// This is used to report the shape of the API call to Prometheus.
+type ShapeBuilder func() APIShapeLabels
+
+// ObserveShapeLatency observes the latency of the API call with the given shape.
+func ObserveShapeLatency(ctx context.Context, methodName string, shape APIShapeLabels, duration time.Duration) {
+	labels := make([]string, len(allLabels)+1)
+	labels[0] = methodName
+
+	for index, label := range allLabels {
+		shapeValue, ok := shape[APIShapeLabel(label)]
+		if !ok {
+			continue
+		}
+
+		switch t := shapeValue.(type) {
+		case string:
+			labels[index+1] = t
+
+		case int:
+			labels[index+1] = strconv.Itoa(t)
+
+		case uint32:
+			intValue, err := safecast.ToInt(t)
+			if err != nil {
+				log.Warn().Str("method", methodName).
+					Str("key", label).
+					Str("type", fmt.Sprintf("%T", t)).
+					Msg("ShapeBuilder could not convert int, skipping reporting")
+				return
+			}
+			labels[index+1] = strconv.Itoa(intValue)
+
+		case int64:
+			intValue, err := safecast.ToInt(t)
+			if err != nil {
+				log.Warn().Str("method", methodName).
+					Str("key", label).
+					Str("type", fmt.Sprintf("%T", t)).
+					Msg("ShapeBuilder could not convert int, skipping reporting")
+				return
+			}
+			labels[index+1] = strconv.Itoa(intValue)
+
+		default:
+			spiceerrors.DebugAssert(func() bool { return false }, "unsupported type %T", t)
+			log.Warn().Str("method", methodName).
+				Str("key", label).
+				Str("type", fmt.Sprintf("%T", t)).
+				Msg("ShapeBuilder returned an unsupported type, skipping reporting")
+			return
+		}
+	}
+
+	o := APIShapeLatency.WithLabelValues(labels...)
+	if exemplarObserver, ok := o.(prometheus.ExemplarObserver); ok {
```

Comment:
Is there a way to type assert this into a global var, instead of keeping the base interface?

Reply:
Not in a nice way :/